File: examples/print.cpp
#include "example_common.h"
#include "iroha/iroha.h"

using namespace iroha;

IDesign *build_design() {
  IDesign *design = new IDesign;
  IModule *module = new IModule(design, "M");
  design->modules_.push_back(module);
  ITable *table = new ITable(module);
  module->tables_.push_back(table);
  IResource *pr = DesignTool::GetOneResource(table, resource::kPrint);
  IResource *as = DesignTool::GetOneResource(table, resource::kAssert);
  IState *st1 = new IState(table);
  IState *st2 = new IState(table);
  table->states_.push_back(st1);
  table->states_.push_back(st2);
  IInsn *pr_insn = new IInsn(pr);
  IRegister *src = DesignTool::AllocConstNum(table, 32, 123);
  pr_insn->inputs_.push_back(src);
  IInsn *as_insn = new IInsn(as);
  IRegister *t = DesignTool::AllocConstNum(table, 0, 1);
  as_insn->inputs_.push_back(t);
  st1->insns_.push_back(pr_insn);
  st1->insns_.push_back(as_insn);
  table->SetInitialState(st1);
  DesignTool::AddNextState(st1, st2);
  DesignTool::Validate(design);

  return design;
}

File: examples/print.py
import sys
sys.path.append('../py')

from iroha import *
from iroha.iroha import *

d = IDesign()
mod = IModule(d, "mod")
tab = ITable(mod)
st1 = IState(tab)
st2 = IState(tab)
tab.states.append(st1)
tab.states.append(st2)
tab.initialSt = st1
design_tool.AddNextState(st1, st2)

r1 = IRegister(tab, "r1")
r1.SetInitialValue(1)
r1.SetType(False, 0)
r123 = IRegister(tab, "r123")
r123.SetInitialValue(123)

design_tool.ValidateIds(d)

print_res = design_tool.GetResource(tab, "print")
print_insn = IInsn(print_res)
print_insn.inputs.append(r123)
st1.insns.append(print_insn)

assert_res = design_tool.GetResource(tab, "assert")
assert_insn = IInsn(assert_res)
assert_insn.inputs.append(r1)
st1.insns.append(assert_insn)

design_tool.ValidateIds(d)

w = DesignWriter(d)
w.Write()
